pkg/tss/tpm_linux.go

```diff
@@ -0,0 +1,252 @@
+	1096043852: "Atmel",
+	1112687437: "Broadcom",
+	1229081856: "IBM",
+	1213220096: "HPE",
+	1297303124: "Microsoft",
+	1229346816: "Infineon",
+	1229870147: "Intel",
+	1279610368: "Lenovo",
+	1314082080: "National Semiconductor",
+	1314150912: "Nationz",
+	1314145024: "Nuvoton Technology",
+	1363365709: "Qualcomm",
+	1397576515: "SMSC",
+	1398033696: "ST Microelectronics",
+	1397576526: "Samsung",
+	1397641984: "Sinosun",
+	1415073280: "Texas Instruments",
+	1464156928: "Winbond",
+	1380926275: "Fuzhou Rockchip",
+	1196379975: "Google",
+}
+
+// TPM versions
+const (
+	TPMVersionAgnostic TPMVersion = iota
+	TPMVersion12
+	TPMVersion20
+)
+
+// TPM interfaces
+const (
+	TPMInterfaceDirect TPMInterface = iota
+	TPMInterfaceKernelManaged
+	TPMInterfaceDaemonManaged
+)
+
+// TPM interfaces with a TPM device on the system.
+type TPM struct {
+	version TPMVersion
+	interf  TPMInterface
+
+	sysPath string
+	rwc     io.ReadWriteCloser
+}
+
+// probedTPM identifies a TPM device on the system, which
+// is a candidate for being used.
+type probedTPM struct {
+	Version TPMVersion
+	Path    string
+}
+
+// TPMInfo contains information about the version & interface
+// of an open TPM.
+type TPMInfo struct {
+	Version      TPMVersion
+	Interface    TPMInterface
+	VendorInfo   string
+	Manufacturer TCGVendorID
+
+	// FirmwareVersionMajor and FirmwareVersionMinor describe
+	// the firmware version of the TPM, but are only available
+	// for TPM 2.0 devices.
+	FirmwareVersionMajor int
+	FirmwareVersionMinor int
+}
+
+func readTPM12VendorAttributes(rwc io.ReadWriter) (TPMInfo, error) {
+	var vendorInfo string
+
+	_, err := tpm.GetManufacturer(rwc)
+	if err != nil {
+		return TPMInfo{}, err
+	}
+
+	return TPMInfo{
+		VendorInfo:           strings.Trim(vendorInfo, "\x00"), // Stubbed
+		Manufacturer:         TCGVendorID(uint32(0)),           // Stubbed
+		FirmwareVersionMajor: int(0),                           // Stubbed
+		FirmwareVersionMinor: int(0),                           // Stubbed
+	}, nil
+}
+
+func readTPM20VendorAttributes(rwc io.ReadWriter) (TPMInfo, error) {
+	var vendorInfo string
+	// The Vendor String is split up into 4 sections of 4 bytes,
+	// for a maximum length of 16 octets of ASCII text. We iterate
+	// through the 4 indexes to get all 16 bytes & construct vendorInfo.
+	// See: TPM_PT_VENDOR_STRING_1 in TPM 2.0 Structures reference.
+	for i := 0; i < 4; i++ {
+		caps, _, err := tpm2.GetCapability(rwc, tpm2.CapabilityTPMProperties, 1, tpmPtVendorString+uint32(i))
+		if err != nil {
+			return TPMInfo{}, fmt.Errorf("tpm2.GetCapability(PT_VENDOR_STRING_%d) failed: %v", i+1, err)
+		}
+		subset, ok := caps[0].(tpm2.TaggedProperty)
+		if !ok {
+			return TPMInfo{}, fmt.Errorf("got capability of type %T, want tpm2.TaggedProperty", caps[0])
+		}
+		// Reconstruct the 4 ASCII octets from the uint32 value.
+		vendorInfo += string(subset.Value&0xFF000000) + string(subset.Value&0xFF0000) + string(subset.Value&0xFF00) + string(subset.Value&0xFF)
```

Comment:
i'm sure confused by this. 
subset.Value&0xFF000000
is not an octet. what does this intend to do?

func main() {
	v := uint32('a' << 24)
	fmt.Println(string(v & 0xFF000000))
}